docs/design.md

```diff
@@ -0,0 +1,785 @@
+Substitution is applied at each layer after all merge actions occur.
+
+Concrete (non-abstract) documents can be used as a source of substitution
+into other documents. This substitution is layer-independent, so given the 3
+layer example above, which includes `global`, `region` and `site` layers, a
+document in the `region` layer could insert data from a document in the
+`site` layer.
+
+Here is a sample set of documents demonstrating subistution:
+
+```yaml
+---
+schema: deckhand/Certificate/v1
+metadata:
+  name: example-cert
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: site
+data: |
+  CERTIFICATE DATA
+---
+schema: deckhand/CertificateKey/v1
+metadata:
+  name: example-key
+  storagePolicy: encrypted
+  layeringDefinition:
+    layer: site
+data: |
+  KEY DATA
+---
+schema: deckhand/Passphrase/v1
+metadata:
+  name: example-password
+  storagePolicy: encrypted
+  layeringDefinition:
+    layer: site
+data: my-secret-password
+---
+schema: armada/Chart/v1
+metadata:
+  name: example-chart-01
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: region
+  substitutions:
+    - dest:
+        path: .chart.values.tls.certificate
+      src:
+        schema: deckhand/Certificate/v1
+        name: example-cert
+        path: .
+    - dest:
+        path: .chart.values.tls.key
+      src:
+        schema: deckhand/CertificateKey/v1
+        name: example-key
+        path: .
+    - dest:
+        path: .chart.values.some_url
+        pattern: INSERT_[A-Z]+_HERE
+      src:
+        schema: deckhand/Passphrase/v1
+        name: example-password
+        path: .
+data:
+  chart:
+    details:
+      data: here
+    values:
+      some_url: http://admin:INSERT_PASSWORD_HERE@service-name:8080/v1
+...
+```
+
+The rendered document will look like:
+
+```yaml
+---
+schema: armada/Chart/v1
+metadata:
+  name: example-chart-01
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: region
+  substitutions:
+    - dest:
+        path: .chart.values.tls.certificate
+      src:
+        schema: deckhand/Certificate/v1
+        name: example-cert
+        path: .
+    - dest:
+        path: .chart.values.tls.key
+      src:
+        schema: deckhand/CertificateKey/v1
+        name: example-key
+        path: .
+    - dest:
+        path: .chart.values.some_url
+        pattern: INSERT_[A-Z]+_HERE
+      src:
+        schema: deckhand/Passphrase/v1
+        name: example-password
+        path: .
+data:
+  chart:
+    details:
+      data: here
+    values:
+      some_url: http://admin:my-secret-password@service-name:8080/v1
+      tls:
+        certificate: |
+          CERTIFICATE DATA
+        key: |
+          KEY DATA
+...
+```
+
+### Control Documents
+
+Control documents (documents which have `metadata.schema=metadata/Control/v1`),
+are special, and are used to control the behavior of Deckhand at runtime.  Only
+the following types of control documents are allowed.
+
+#### DataSchema
```

Comment:
Good question!

So, ` DataSchema` is literally the jsonschema schema used to validate documents of the type its `metadata.name` matches.

Re: how to validate them, it will be the schema defined by `deckhand/DataSchema/v1`, which will essentially by the spec-standard jsonschema meta schema.

* [Here](http://json-schema.org/documentation.html) is the doc page for json schema
* [Here](http://json-schema.org/schema) is the direct link for the meta-schema